### PR TITLE
New hackathon-related emails

### PR DESCRIPTION
**Pull request by porton2**
##### Description

The job described in https://github.com/gitcoinco/web/issues/6559
Also sends a email when a new hackathon bounty is created.

##### Refers/Fixes

Fixes: https://github.com/gitcoinco/web/issues/6559

##### Testing

The following URLs are used for testing.
```python
    path(
        '_administration/email/hackathon_ends/<str:hackathon>/<str:username>',
        retail.emails.hackathon_ends,
        name='hackathon_ends'),
    path(
        '_administration/email/bounty_not_submitted/<int:bounty>',
        retail.emails.bounty_not_submitted,
        name='bounty_not_submitted'),
    path(
        '_administration/email/bounty_added_to_event/<int:bounty>/<str:profile>',
        retail.emails.bounty_added_to_event,
        name='bounty_added_to_event'),
```

I didn't test sending emails, because SendGrid does not work for me in an unknown reason. However the tests of the above URLs together with mail logs show that emails are correct. Please test in your setting containing working SendGrid, to be sure.

### CODE REVIEW

**Comment by porton2** [comment]
Please review https://github.com/vporton/web/blob/hackathon-emails/app/dashboard/management/commands/hackaton_daily_emails.py#L38 - L45.

The code seems correct, but only someone with SendGrid working can test it, to be completely sure it does not send excessive emails.

**Comment by porton2** [comment]
I forgot a small detail, editing email suppressions. I've added it to the PR now.

File: app/app/urls.py [hunk 1]
```diff
@@ -514,6 +514,21 @@
...
+        retail.emails.hackathon_ends,
+        name='hackathon_ends'
+    ),
+    path(
+        '_administration/email/bounty_not_submitted/<int:bounty>',
+        retail.emails.bounty_not_submitted,
+        name='bounty_not_submitted'
+    ),
+    path(
+        '_administration/email/bounty_added_to_event/<int:bounty>/<str:profile>',
```
Comment (thelostone-mc) [inline app/app/urls.py:528]: ^ could we update URLs to `_administration/email/hackathons/*`
All the URLs

The other we'll have to group off them slowly to make it a li more maintainable

File: app/dashboard/templates/emails/hackathons/bounty_not_submitted.html [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+{% endcomment %}
+{% load i18n %}
+
+{% block content %}
+<h1>{% trans "Bounty not submitted" %}</h1>
+
+{% with title=bounty.title url=bounty.get_absolute_url %}
+<p style="margin-bottom:0.25rem;">
+  {% blocktrans %}The bounty <b><a href="{{ url }}">{{ title }}</a></b> <em>not submitted by you</em> ends in 4-5 days!{% endblocktrans %}
```
Comment (thelostone-mc) [inline app/dashboard/templates/emails/hackathons/bounty_not_submitted.html:27]: Shouldn't the hackathon name be mentioned here ?

File: app/app/urls.py [hunk 3]
```diff
@@ -514,6 +514,21 @@
         retail.emails.successful_contribution,
         name='admin_successful_contribution'
     ),
+    path(
+        '_administration/email/hackathon_ends/<str:hackathon>/<str:username>',
+        retail.emails.hackathon_ends,
+        name='hackathon_ends'
+    ),
+    path(
+        '_administration/email/bounty_not_submitted/<int:bounty>',
```
Comment (thelostone-mc) [inline app/app/urls.py:523]: Does this not need hackathon id ? Is that taken from bounty

File: app/dashboard/management/commands/hackaton_daily_emails.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+'''
```
Comment (thelostone-mc) [inline app/dashboard/management/commands/hackaton_daily_emails.py:1]: ^ file name needs to corrected

File: app/dashboard/management/commands/hackaton_daily_emails.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+from django.conf import settings
+from django.core.management.base import BaseCommand
+from django.utils import timezone
+
+from dashboard.models import Activity, Bounty, BountyFulfillment, Profile
+from marketing.mails import bounty_feedback, bounty_not_submitted
+
+
+class Command(BaseCommand):
+    help = 'sends more timed hackathon emails'
```
Comment (thelostone-mc) [inline app/dashboard/management/commands/hackaton_daily_emails.py:29]: description needs to be more descriptive !

File: app/dashboard/management/commands/hackaton_daily_emails.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    def handle(self, *args, **options):
+        if settings.DEBUG:
+            print("not active in non prod environments")
+            return
+
+        # TODO: Probably not the most efficient way, need instead use self-join.
+        # However once per day only for users started work on a bounty not extremely slow.
```
Comment (thelostone-mc) [inline app/dashboard/management/commands/hackaton_daily_emails.py:37]: Yeah this seems to be a bit of an overhaul.
@danlipert what do you think ?

File: app/dashboard/helpers.py [hunk 7]
```diff
@@ -557,6 +557,11 @@ def create_new_bounty(old_bounties, bounty_payload, bounty_details, bounty_id):
             logger.error(f'{e} encountered during new bounty creation for: {url}')
             new_bounty = None
 
+    if new_bounty.event:
+        for registration in HackathonRegistration.objects.filter(hackathon=new_bounty.event):
+            # TODO: Run this in background:
```
Comment (thelostone-mc) [inline app/dashboard/helpers.py:562]: Do we still need this todo / has the cronjob been set ?

**Comment by vporton** [comment]
> @porton2 could you add
> 
>     * screenshots on mobile + desktop

Desktop screenshot:

Screenshot from 2020-05-04 08-04-49
Screenshot from 2020-05-04 08-04-55
Screenshot from 2020-05-04 08-04-56

Mobile screenshot look like:

Screenshot from 2020-05-04 08-07-17
Screenshot from 2020-05-04 08-07-18
Screenshot from 2020-05-04 08-07-20

>     * test it out with an email client (how it views in _admin might be a lil different than )

I told that SendGrid does not work for me. To test it with an email client I would need to write my own email delivery function. I think it's better if somebody with working SendGrid tests.

File: app/app/urls.py (hunk 1 above)
Comment (vporton) [reply app/app/urls.py]: Done.

File: app/dashboard/templates/emails/hackathons/bounty_not_submitted.html (hunk 2 above)
Comment (vporton) [reply app/dashboard/templates/emails/hackathons/bounty_not_submitted.html]: Done.

File: app/app/urls.py (hunk 3 above)
Comment (vporton) [reply app/app/urls.py]: Taken from the bounty.

File: app/dashboard/management/commands/hackaton_daily_emails.py (hunk 4 above)
Comment (vporton) [reply app/dashboard/management/commands/hackaton_daily_emails.py]: Done.

File: app/dashboard/management/commands/hackaton_daily_emails.py (hunk 5 above)
Comment (vporton) [reply app/dashboard/management/commands/hackaton_daily_emails.py]: Done.

File: app/dashboard/helpers.py (hunk 7 above)
Comment (vporton) [reply app/dashboard/helpers.py:562]: The cronjob has been set, but to test it needs working SendGrid and it does not work for me (despite I added my email to their "owner" emails list). Somebody test it.

**Comment by frankchen07** [comment]
hey @vporton 

just wanted to confirm that the functionality you coded up covers:

1. ability to send emails to a very targeted group of people (people on X hackathon)

2. ability to send emails to a very targeted group of people performed certain actions (people on X hackathon who started work but haven't yet submitted)

3. ability to auto send an email to a targeted group of people based on another data model (email sent to people on X hackathon when only 48 hours remain before hackathon end time)

From the looks of it, you've coded up the 48 hours till end, bounty not submitted, and also bounty added to a hackathon emails, but I'm wondering how extensible it is. 

1. Can we dynamically select which hackathon we send to (if hackathon Y is passed, and we want to send for hackathon X), or is it a code change in the `.py` file?

2. Basically, if some of our non-technical team members instead want to send emails to a slightly different set of folks on a hackathon and maybe part of a project, is it possible to generalize it in a way where one of us can go in and switch that up, or would we have to go through the entire process of creating a new html template, adding the url and coding up the parameters we want? I think right now it would be a modification of creating a new method in `mails.py`

In a separate ticket, I'll probably start another bounty where we can add the ability to unsubscribe to hackathon mailing lists under the profile settings in gitcoin.co/settings. That is out of scope on this bounty, just fyi as a side thought / follow up bounty

**Comment by vporton** [comment]
> hey @vporton, really appreciate the fast PR. Thanks for jumping on this!
> 
> just wanted to confirm that the functionality you coded up covers:
> 
>     1. ability to send emails to a very targeted group of people (people on X hackathon)

Yes.

>     2. ability to send emails to a very targeted group of people performed certain actions (people on X hackathon who started work but haven't yet submitted)

Yes.

>     3. ability to auto send an email to a targeted group of people based on another data model (email sent to people on X hackathon when only 48 hours remain before hackathon end time)

Yes.

> From the looks of it, you've coded up the 48 hours till end, bounty not submitted, and also bounty added to a hackathon emails, but I'm wondering how extensible it is.
> 
>     1. Can we dynamically select which hackathon we send to (if hackathon Y is passed, and we want to send for hackathon X), or is it a code change in the `.py` file?

I don't understand.

It sends for all hackathons for which a user is related.

>     2. Basically, if some of our non-technical team members instead want to send emails to a slightly different set of folks on a hackathon and maybe part of a project, is it possible to generalize it in a way where one of us can go in and switch that up, or would we have to go through the entire process of creating a new html template, adding the url and coding up the parameters we want? I think right now it would be a modification of creating a new method in `mails.py`

I don't understand you: You ask whether it is possible to change the logic without changing the code. But I don't understand which kind of logic change you mean.

**Comment by frankchen07** [comment]
@vporton - thanks for the clarifications.

> 48 hrs until the end

I see what you're saying. The logic coded here would send an email for all hackathon registrants who are part of an hackathon ending in 48 hours, whether it be multiple or only 1 hackathon.

> is it possible to change the logic without changing the code

I think this is essentially what I asked, and the answer is probably no. As an example these two use cases:

email 1 (which you built out): hey [user], I see you're part of [hackathon X] and you've started [project Y], don't forget to submit!

email 2 (not built out): hey [user], [hackathon Z] is coming up and because you were part of [hackathon X], you should check it out!

email 2 has to be coded up separately because it has completely different logic.

The question that I think I originally wanted to ask was if there was a friendly UI solution to changing the logic of an email, but I think I can answer that - no, we'll have to code up separate use cases. 

let me ask our hackathon ops lead to make sure we're not missing out on any other priority use cases, if there are any additional, I'll note them here (and also note that may increase the bounty cost as well).

**Comment by frankchen07** [comment]
@willsputra - is it possible to get your design expertise on some of these HTMLs?

**Comment by frankchen07** [comment]
hi @vporton - thanks for being patient. For now, I think those 3 use cases are good.

@willsputra generously made these designs:

is it possible to update the htmls to these and ensure they're dynamic?:

1 - 48 hours
2 - Bounty not submitted
3 - Bounty added

**Comment by vporton** [comment]
"is it possible to update the htmls to these and ensure they're dynamic?"

What do you mean by "dynamic"?

I may try but I am not an expert in making good looking HTML.

**Comment by frankchen07** [comment]
I should be more clear - dynamic meaning "new bounties added to [HACKATHON]" - not just hardcoding "new york blockchain week", but I think you understand that already. 

I'm just using non-standard terms to describe it.

**Comment by alexvotofuture** [comment]
Hey all, from the ops team perspective the templates, use cases (except the "new bounty added to hackathon" one), and approach presented here look good to get us started!

Ultimately we'll want to be able to construct queries ourselves (or at least pick from some common ones Frank sets us up with), use a WYSIWYG email designer (for example, SendGrid has one that constructs dynamic html templates), and combine those to send an email either once or on a recurring/automatic/triggered basis. This will save us having to reach out to @frankchen07 for db queries and @willsputra for html designs each time. I can share the full proposed lists of automated emails for hackers and sponsors via Slack, if it's helpful for identifying the kinds of queries we'd run. 

We may eventually consider giving this power to people who self-host hackathons using our platform.

But I'd say for now let's push forward with this and measure how effective it is at getting hackers engaged!

**Comment by frankchen07** [comment]
@vporton - can we add one more use case?

4. Ability to send an email to hackers who registered for a hackathon, but haven't done anything (no start works, no work submitted).

We can use a repurposed HTML off of this one: 

1 - 48 hours

I understand this bounty has increased in scope, so I've increased the bounty amount by 50%. :)

**Comment by vporton** [comment]
> @vporton - can we add one more use case?
> 
>     1. Ability to send an email to hackers who registered for a hackathon, but haven't done anything (no start works, no work submitted).

It looks meaningless for me: If the hacker has not started work and it remains just 8 hours, he is not interested, he won't do anything anyway (most likely). Why to notify him?

**Comment by vporton** [comment]
> > @vporton - can we add one more use case?
> > 
> > 1. Ability to send an email to hackers who registered for a hackathon, but haven't done anything (no start works, no work submitted).
> 
> It looks meaningless for me: If the hacker has not started work and it remains just 8 hours, he is not interested, he won't do anything anyway (most likely). Why to notify him?

So, to do this or not to do? Somebody authoritative (and I know he is authoritative) make the decision. Owocki?

**Comment by frankchen07** [comment]
@vporton - I see your point. This coincides with some of our thinking that not everyone who joins a hackathon might be there to hack. They might be there for alternative resources, networking, doing quests, and attending workshops.

As we mature our hackathon offerings we'll revisit this to see what form these emails take.

I guess from my standpoint, if we do need this use case, I can easily do it manually using the data in our DB, if it comes to that. 

In that case we can skip 4, and just make sure we get the three use cases above into production.

**Comment by frankchen07** [comment]
hey @vporton - are these comments above and [here](https://github.com/gitcoinco/web/issues/6559#issuecomment-626835343) sufficient for you?

**Comment by vporton** [comment]
> hey @vporton - are these comments above and [here](https://github.com/gitcoinco/web/issues/6559#issuecomment-626835343) sufficient for you?

It seems I finished my work on the issue and as I understand I we decided not to do:

> Ability to send an email to hackers who registered for a hackathon, but haven't done anything (no start works, no work submitted).

So, the work (except of designing the templates) is finished. Right?

**Comment by frankchen07** [comment]
Yes. Just wanted to make sure the HTML templates were finished as well.

**Comment by thelostone-mc** [comment]
@vporton could you share the screenshots as and when changes are made -> it would make it easier to validate for folks reviewing the PR.
We'd be able to get it in quicker that way

**Comment by frankchen07** [comment]
heyo @vporton, just wanted to follow up on the screenshots

**Comment by vporton** [comment]
> heyo @vporton, just wanted to follow up on the screenshots

It seems we agreed it will be done by CSS experts. I can try, but don't warrant quality.

**Comment by frankchen07** [comment]
@vporton - sorry that was probably lost in communication. 

above you said: "I may try but I am not an expert in making good looking HTML."

I assumed that meant you couldn't guarantee quality but you'd give it a shot, so I thought you were working.

Hopefully we can get the HTML in by this week?

**Comment by vporton** [comment]
> @vporton - sorry that was probably lost in communication.
> 
> above you said: "I may try but I am not an expert in making good looking HTML."
> 
> I assumed that meant you couldn't guarantee quality but you'd give it a shot, so I thought you were working.
> 
> Hopefully we can get the HTML in by this week?

Sorry, I can't. Let a master person do it, not me.

**Comment by frankchen07** [comment]
@vporton - got it. Thanks for letting me know. I'll see if I can dig someone up. 

I'll make sure to pay you fairly for the work you've done once I close this issue out.

**Comment by danlipert** [comment]
@frankchen07 I'm going to close this for now - let me know when/if you want to get this finished out

**Comment by frankchen07** [comment]
@vporton - paid you out. Circumstances have changed and it may or may not be used, but fair is fair and I appreciate your jumping on on the issue so quickly. Let me know if you got that payment and definitely will think of you for future bounties!

**Comment by vporton** [comment]
> Let me know if you got that payment

I've received.

**Pull request closed without merging**